### ISSUE TITLE
[aws-s3-deployment] support for empty folder deployement

### ISSUE THREAD

**Issue by BenoitDel** [aws/aws-cdk#9666]
### Description

With the following assets folder:

assets/
   |----> FolderA/ 

Instantiate BucketDeployement has no effect.
```
new BucketDeployment(this, "DeployWebsite", {
      sources: [Source.asset("./assets")],
      destinationBucket: myBucket
    }); // no effect
```
What would be expected?

A bucket with the following structure:

Mybucket:
   |----> FolderA/ 

### Use Case

required to generate folder-level access bucket policies

### Proposed Solution

As a workaround, i use empty file to force BucketDeployement to render the folder structures
And delete the empty file afterwards

assets/
   |----> FolderA/ ---->emptyFiles.txt 

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by jonlegendre** [comment]
I am working around this limitation with the following code:

    private ProvisionS3Folder(targetBucket: s3.IBucket, targetFolder: string, policy: AwsCustomResourcePolicy): AwsCustomResource {
      //NOTE once this issue is fixed we can use a better command for this instead of a custom: https://github.com/aws/aws-cdk/issues/9666
      var folderName = targetFolder.endsWith('/') || targetFolder.endsWith('\\') ? targetFolder : `${targetFolder}/`;
      var folder = new AwsCustomResource(this, `provision_s3_folder_${targetFolder}_custom`, {
        onCreate: {
          service: 'S3',
          action: "putObject",
          parameters: { 
            Bucket: targetBucket.bucketName,
            Key: folderName,
          },
          physicalResourceId: { id: `provision_s3_folder_${folderName}` },
        },
        policy: policy,
      });
      folder.node.addDependency(targetBucket);
      return folder;
    }
I assume that similar code can be used in the CDK source code itself. I will try to take a peek when I am off work to see if this is something I can address via a pull request. 
I am currently deleting the bucket on delete, and don't need the folder recreated on update so I haven't implemented those above, but they are easy enough to add if needed by following the pattern above.

**Comment by senstb** [comment]
I recently had an ask similar to this and was able to come up with a CDK native workaround to this using the ['destinationKeyPrefix'](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-s3-deployment-readme.html#objects-metadata) option in the BucketDeployment props.

You can setup your project source like what can be seen below, you'll need a folder/file for git management but the '_create' file can be empty 

```
projectRoot/
|-->bucket_config
|---->_create
|-->lib
|---->resource_stack.ts
```

In the 'resource_stack.ts' stack definition you can define something similar to what can be seen below 

```
export class MyResourceStack extends DeploymentStack {
  constructor(scope: App, id: string, props: DeploymentStackProps) {
    super(scope, id, props);

    // S3 bucket
    const S3BucketExample = new Bucket(this, 'S3ExampleBucket',{
        bucketName: 's3-cdk-example',
        publicReadAccess: false,
        removalPolicy: RemovalPolicy.DESTROY
    })

    // S3 Bucket Prefix
    const archivePrefixCreation = new s3deploy.BucketDeployment(this, 'ArchivePrefixDeployment',{
      sources:[s3deploy.Source.asset('./bucket_config')],
      destinationBucket: S3BucketExample,
      destinationKeyPrefix: 'Archive/'
    })
```

This will leverage an empty '_create' file stored locally and upload the object and create the 'Archive/' prefix within your S3 bucket, resulting in a configuration similar to what's below. 

```
s3://s3-cdk-example/
|-->Archive/
|---->_create
```

One general limitation with the creation of 'folders' in S3 is that S3 doesn't necessarily store things like a standard drive; since it's object-based storage all a 'folder' is on the S3 side is basically just a [prefix pointer to help tag/logically sort objects stored within a given bucket](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html).

I do think having a method in the [aws-s3-deployment](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-s3-deployment-readme.html#supported-sources) module to just create a prefix would be useful, like an 'aws-s3-deployment.BucketPrefix' class modeled off the 'aws-s3-deployment.BucketDeployment' class to generate the prefix/folder in the target S3 bucket.

**Comment by 094459** [comment]
I am using aws-s3-deployment to create an "empty" folder using the following code:

```
        s3deploy.BucketDeployment(self, "File",
        sources=[s3deploy.Source.asset("./files")],
        destination_bucket=redshift_bucket,
        destination_key_prefix="files/",
        prune=False,
        retain_on_delete=False
        )
```
In my CDK folder I have a folder called "files" with a dummy file in it. Whilst this does get created, and it looks at first glance ok, because of the way S3 works, when I use the aws cli to inspect I can see the following output:

```
$ aws s3 ls s3://mwaa-redshift-blog/files/ --summarize
2021-05-17 11:32:57          6 readme.txt
Total Objects: 1
```

When I do the same thing, using the AWS console to create the folder and upload the file I get the following:

```
$ aws s3 ls s3://mwaa-redshift-blog/files/ --summarize
2021-05-17 11:32:57          6 readme.txt
Total Objects: 2
```
Whilst this is the expected behaviour from my understanding, what I am puzzled is why creating the folder/uploading the file results in a different outcome for doing the same thing in effect.

The application I am using fails to detect the folder when it is created using CDK, but works when it is created manually.

**Comment by fivepapertigers** [comment]
> S3 has no concept of folders though. S3 is an object store. Each object is a blob, which is identified by a key. A folder is not an object or a blob. It's a filesystem concept.

This is true, except that "folder objects" are meaningful in filesystem services that integrate with S3, such as Transfer Family. I came to this issue because my team is specifically interested in pre-provisioning folders at deploy-time on a Transfer Family SFTP server that is backed by S3. The way to do this is to create an empty object with a trailing slash, but this bug prevents that. The `.keep` workaround is simple enough and a reasonable solution, but it comes at the expense of explaining the presence/significance of that file to the consuming parties of our SFTP server.

So yes, folders are a bit of a hack/illusion, but in the case of filesystem integrations that are built on top of S3, it's a somewhat of a critical one.

**Comment by moltar** [comment]
> I came to this issue because my team is specifically interested in pre-provisioning folders at deploy-time on a Transfer Family SFTP server that is backed by S3. 

Ah, interesting. Thanks for sharing. Didn't know this.

**Comment by moltar** [comment]
I think another problem with empty folders, unrelated to AWS, is that git does not support empty folders too. So if you manage your whole setup as code, then you'll need a hack for git (e.g. `.gitkeep` file) and then you also have to tell Source _not_ to use that file. So, overall, seems like too many complications.

If all you need is some pre-defined keys in S3, maybe a custom resource would be an easier solution.
